File: src/main/scala/com/flink/tutorials/scala/utils/taobao/UserBehaviorSource.scala
package com.flink.tutorials.scala.utils.taobao

import java.io.InputStream

import org.apache.flink.streaming.api.functions.source.RichSourceFunction
import org.apache.flink.streaming.api.functions.source.SourceFunction.SourceContext

class UserBehaviorSource(path: String) extends RichSourceFunction[UserBehavior] {

  var isRunning: Boolean = true
  // 输入源
  var streamSource: InputStream = _

  override def run(sourceContext: SourceContext[UserBehavior]): Unit = {
    // 从项目的resources目录获取输入
    streamSource = this.getClass.getClassLoader.getResourceAsStream(path)
    val lines: Iterator[String] = scala.io.Source.fromInputStream(streamSource).getLines
    var isFirstLine: Boolean = true
    var timeDiff: Long = 0
    var lastEventTs: Long = 0
    while (isRunning && lines.hasNext) {
      val line = lines.next()
      val itemStrArr = line.split(",")
      val eventTs: Long = itemStrArr(4).toLong
      if (isFirstLine) {
        // 从第一行数据提取时间戳
        lastEventTs = eventTs
        isFirstLine = false
      }
      val userBehavior = UserBehavior(itemStrArr(0).toLong, itemStrArr(1).toLong, itemStrArr(2).toInt, itemStrArr(3), eventTs)
      // 输入文件中的时间戳是从小到大排列的
      // 新读入的行如果比上一行大，sleep，这样来模拟一个有时间间隔的输入流
      timeDiff = eventTs - lastEventTs
      if (timeDiff > 0)
        Thread.sleep(timeDiff * 1000)
      sourceContext.collect(userBehavior)
      lastEventTs = eventTs
    }
  }

  override def cancel(): Unit = {
    streamSource.close()
    isRunning = false
  }
}

File: src/main/java/com/flink/tutorials/java/utils/taobao/UserBehaviorSource.java
package com.flink.tutorials.java.utils.taobao;

import org.apache.flink.streaming.api.functions.source.SourceFunction;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;

public class UserBehaviorSource implements SourceFunction<UserBehavior> {

    private boolean isRunning = true;
    private String path;
    private InputStream streamSource;

    public UserBehaviorSource(String path) {
        this.path = path;
    }

    @Override
    public void run(SourceContext<UserBehavior> sourceContext) throws Exception {
        // 从项目的resources目录获取输入
        streamSource = this.getClass().getClassLoader().getResourceAsStream(path);
        BufferedReader br = new BufferedReader(new InputStreamReader(streamSource));
        String line;
        boolean isFirstLine = true;
        long timeDiff = 0;
        long lastEventTs = 0;
        while (isRunning && (line = br.readLine()) != null) {
            String[] itemStrArr = line.split(",");
            long eventTs = Long.parseLong(itemStrArr[4]);
            if (isFirstLine) {
                // 从第一行数据提取时间戳
                lastEventTs = eventTs;
                isFirstLine = false;
            }
            UserBehavior userBehavior = UserBehavior.of(Long.parseLong(itemStrArr[0]),
                    Long.parseLong(itemStrArr[1]), Integer.parseInt(itemStrArr[2]),
                    itemStrArr[3], eventTs);
            // 输入文件中的时间戳是从小到大排列的
            // 新读入的行如果比上一行大，sleep，这样来模拟一个有时间间隔的输入流
            timeDiff = eventTs - lastEventTs;
            if (timeDiff > 0)
                Thread.sleep(timeDiff * 1000);
            sourceContext.collect(userBehavior);
            lastEventTs = eventTs;
        }
    }

    // 停止发送数据
    @Override
    public void cancel() {
        try {
            streamSource.close();
        } catch (Exception e) {
            System.out.println(e.toString());
        }
        isRunning = false;
    }

}
